sled-agent/src/bootstrap/agent.rs

```diff
@@ -729,6 +750,163 @@ impl Agent {
+            .map_err(BootstrapError::Cleanup)?;
+        sled_hardware::cleanup::delete_omicron_vnics(&self.log)
+            .await
+            .map_err(BootstrapError::Cleanup)?;
+        illumos_utils::opte::delete_all_xde_devices(&self.log)?;
+        Ok(())
+    }
+
+    async fn uninstall_storage_locked(
+        &self,
+        _state: &tokio::sync::MutexGuard<'_, SledAgentState>,
+    ) -> Result<(), BootstrapError> {
+        let datasets = zfs::get_all_omicron_datasets()
+            .map_err(BootstrapError::ZfsDatasetsList)?;
+        for dataset in &datasets {
+            info!(self.log, "Removing dataset: {dataset}");
+            zfs::Zfs::destroy_dataset(dataset)?;
+        }
+
+        Ok(())
+    }
+
+    /// Resets this sled, removing:
+    ///
+    /// - All control plane zones (except the switch zone)
+    /// - All sled-local configuration
+    /// - All underlay networking
+    /// - All storage managed by the control plane
+    ///
+    /// This API is intended to put the sled into a state where it can
+    /// subsequently be initialized via RSS.
+    pub async fn sled_reset(&self) -> Result<(), BootstrapError> {
+        let mut state = self.sled_state.lock().await;
+
+        if let SledAgentState::After(_) = &mut *state {
+            // We'd like to stop the old sled agent before starting a new
+            // hardware monitor -- however, if we cannot start a new hardware
+            // monitor, the bootstrap agent may be in a degraded state.
+            let server = match std::mem::replace(
+                &mut *state,
+                SledAgentState::Before(None),
+            ) {
+                SledAgentState::After(server) => server,
+                _ => panic!(
+                    "Unexpected state (we should have just matched on it)"
+                ),
+            };
+            server.close().await.map_err(BootstrapError::SledError)?;
+        };
+
+        // Try to reset the sled, but do not exit early on error.
+        let result = async {
+            self.uninstall_zones_locked(&state).await?;
+            self.uninstall_sled_local_config_locked(&state).await?;
+            self.uninstall_networking_locked(&state).await?;
+            self.uninstall_storage_locked(&state).await?;
+            Ok::<(), BootstrapError>(())
+        }
+        .await;
+
+        // Try to restart the bootstrap agent hardware monitor before
+        // returning any errors from reset.
+        match &mut *state {
+            SledAgentState::Before(None) => {
+                let hardware_monitor = self.start_hardware_monitor()
+                    .await
+                    .map_err(|err| {
+                        warn!(self.log, "Failed to restart bootstrap agent hardware monitor");
+                        err
+                    })?;
+                *state = SledAgentState::Before(Some(hardware_monitor));
+            }
+            _ => {}
```

Comment:
Just double-checking my own understanding: at this point, we cannot be in `SledAgentState::After()` (because of the swap up on line 867). We could be already be in `Before(Some(_))`, in which case we do nothing, and if we were in `Before(None)` we're now `Before(Some(_))` instead thanks to line 899?

Reply:
Yup! I really wish there was a better way for me to have the `Before` and `After` states without the `Option` in the before state, but unfortunately, starting the hardware monitor *is* a fallible operation, so it's possible to fall-back to that as a defunct state.

In plain English, this function should read as the following:

1. If a Sled Agent server is running, stop it.
2. Try to uninstall everything we can.
3. If the bootstrap agent is not monitoring hardware, for any reason, start the monitor.

Step (3) could trigger because we stopped the sled agent. It could *also* trigger because we previously failed to restart the hardware monitor.